dkist/wcs/models.py

```diff
@@ -0,0 +1,329 @@
+        │  │           │
+        │  │  ┌────────┤
+        │  │  │        │
+        ▼  ▼  ▼        ▼
+      ┌─────────┐  ┌────────┐
+      │Celestial│  │Temporal│
+      └─┬───┬───┘  └───┬────┘
+        │   │          │
+        │   │          │
+        │   │          │
+        ▼   ▼          ▼
+       lon lat       time
+
+    This could trivially be reproduced using `~astropy.modeling.models.Mapping`.
+
+    The complexity is in the reverse transform, where the inverse Celestial
+    transform is also dependent upon the pixel coordinate z.
+    This means that the output of the inverse Temporal transform has to be
+    duplicated as an input to the Celestial transform's inverse.
+    This is achieved by the use of the ``Mapping`` models in
+    ``CoupledCompoundModel.inverse`` to create a multi-stage compound model
+    which duplicates the output of the right hand side model::
+
+       lon lat       time
+        │   │         │
+        │   │         ▼
+        │   │     ┌─────────┐
+        │   │     │Temporal'│
+        │   │     └──┬──┬───┘
+        │   │    z   │  │
+        │   │  ┌─────┘  │
+        │   │  │        │
+        ▼   ▼  ▼        │
+      ┌──────────┐      │
+      │Celestial'│      │
+      └─┬───┬────┘      │
+        │   │           │
+        ▼   ▼           ▼
+        x   y           z
+
+    """
+
+    _separable = False
+
+    def __init__(self, op, left, right, name=None, shared_inputs=1):
+        if op != "&":
+            raise ValueError(
+                f"The {self.__class__.__name__} class should only be used with the & operator."
+            )
+        super().__init__(op, left, right, name=name)
+        self.n_inputs = self.n_inputs - shared_inputs
+        self.inputs = self.inputs[:-shared_inputs]
+        self.shared_inputs = shared_inputs
+
+    def _evaluate(self, *args, **kw):
+        leftval = self.left(*(args[:self.left.n_inputs]), **kw)
+        rightval = self.right(*(args[-self.right.n_inputs:]), **kw)
+
+        return self._apply_operators_to_value_lists(leftval, rightval, **kw)
+
+    @property
+    def inverse(self):
+        left_inverse = self.left.inverse
+        right_inverse = self.right.inverse
+
+        total_inputs = self.n_outputs
+        n_left_only_inputs = total_inputs - self.shared_inputs
+
+        # Pass through arguments to the left model unchanged while computing the right output
+        mapping = list(range(n_left_only_inputs))
+        step1 = m.Mapping(mapping) & right_inverse
+
+        # Now pass through the right outputs unchanged while also feeding them into the left model
+
+        # This mapping duplicates the output of the right inverse to be fed
+        # into the left and also out unmodified at the end of the transform
+        inter_mapping = mapping + list(range(max(mapping) + 1, max(mapping) + 1 + right_inverse.n_outputs)) * 2
+        step2 = m.Mapping(inter_mapping) | (left_inverse & m.Mapping(list(range(right_inverse.n_outputs))))
```

Comment:
Is this `m.Mapping(list(range(right_inverse.n_outputs)))` equivalent to `m.Identity(right_inverse.n_outputs)`?
If so it may improve readability.